File: week6-Python/pset6/credit/credit.py
from cs50 import get_int


def main():
    # Prompt the user for the card number
    card = get_int("Number: ")
    card = str(card)
    
    # Validate the length of the card
    if validate_length(card):
        # if length is ok, apply Luhn's algorithm
        if validate_card(card):
            # Detect the type of the card
            brand_of_card = type_of_card(card)
            # print the type of the card
            print(brand_of_card)
        
    else:
        # If it's not a valid card, print invalid
        print("INVALID")


def validate_length(card):
    # return true if the length is one of the below
    return (len(card) == 13) or (15 <= len(card) <= 16)


def validate_card(card):
    # separate the digits for the Luhn's calculation
    first_digits = card[-2::-2]
    second_digits = card[-1::-2]
    
    first_digits_int = []
    second_digits_int = []
    
    # Calculate the first part of the algorithm
    for digit in first_digits:
        digit = int(digit) * 2
        
        if len(str(digit)) > 1:
            digit = str(digit)
            
            first_digits_int.append(int(digit[0]))
            first_digits_int.append(int(digit[1]))
        
        else:
            first_digits_int.append(digit)
    
    # Calculate the second part of the algorithm
    for digit in second_digits:
        digit = int(digit)
        second_digits_int.append(digit)
    
    # Add the digits and check if it's divisible by 10
    sum_digits = str(sum(first_digits_int) + sum(second_digits_int))
    
    if sum_digits[-1] == '0':
        return True
    
    else:
        return False


def type_of_card(card):
    
    amex_numbers = ['34', '37']
    master_numbers = ['51', '52', '53', '54', '55']
    visa_numbers = ['4']
    
    # Verify the type of the card
    if card[0:2] in amex_numbers:
        return f"AMEX"
        
    elif card[0:2] in master_numbers:
        return f"MASTERCARD"
        
    elif card[0] in visa_numbers:
        return f"VISA"
    
    else:
        return f"INVALID"


main()

File: week1-C/pset1/credit/credit.c
#include <stdio.h>
#include <cs50.h>
#include <math.h>

long get_number(void);
int sum_numbers(long number);
bool validate_length(long number);
void check_type(long number);

int main(void)
{
    //get the number from the user
    long number = get_number();

    //validate the card's length
    if (validate_length(number))
    {   
        //if length is valid than execute the luhn's algorithm
        int result_sum = sum_numbers(number);
        //if the sum ends with 0 check the type, else just print invalid
        if (result_sum % 10 == 0)
        {
            check_type(number);
        }
        
        else
        {
            printf("%s\n", "INVALID");
        }
    }
    
    else
    {
        printf("%s\n", "INVALID");
    }
}


//function for getting the number from user
long get_number(void)
{
    long number;
    
    do
    {
        number = get_long("Number: ");
    }
    while (number <= 0);
    
    return number;
}

//function to perform luhn's algorithm
int sum_numbers(long number)
{
    int r;
    
    long n = number;
    
    int i = 0; 
    int sum = 0;
    
    do
    {
        r = n % 10;
        //see if it's divisible by 2
        if (i % 2 == 0)
        {
            //if it is, just add r to sum
            sum += r;
        }
      
        else
        {
            //if it isn't multiply by 2 and separate digits if necessary
            r = r * 2;
            sum += r / 10 + r % 10;
        }
      
        n = n / 10;
        i++;
    }
    while (n > 0);
    
    return sum;
}

//function to validate the length
bool validate_length(long number)
{
    bool valid;
    
    long n = number;
    
    int i;
    
    for (i = 0; n > 0; i++)
    {
        n = n / 10;
    }
    //return tru if the length is equal to 13, 15 or 16
    if (i == 13 || i == 15 || i == 16)
    {
        valid = true;
    }
    
    
    else
    {
        valid = false;
    }
    
    return valid;
}

//function to check the type
void check_type(long number)
{
    if (number > 340000000000000 && number < 350000000000000)
    {
        printf("%s\n", "AMEX");
    }
    else if (number > 370000000000000 && number < 380000000000000)
    {
        printf("%s\n", "AMEX");
    }
    else if (number > 5100000000000000 && number < 5600000000000000)
    {
        printf("%s\n", "MASTERCARD");
    }
    else if (number > 4000000000000 && number < 5000000000000)
    {
        printf("%s\n", "VISA");
    }
    else if (number > 4000000000000000 && number < 5000000000000000)
    {
        printf("%s\n", "VISA");
    }
    else
    {
        printf("%s\n", "INVALID");
    }
}